File: src/main/java/com/vielheit/graph/config/Neo4jConfig.java
package com.vielheit.graph.config;

import org.neo4j.ogm.session.SessionFactory;
import org.springframework.boot.autoconfigure.jdbc.DataSourceBuilder;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.neo4j.repository.config.EnableNeo4jRepositories;
import org.springframework.data.neo4j.transaction.Neo4jTransactionManager;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.sql.DataSource;

@Configuration
@EnableTransactionManagement
@EnableNeo4jRepositories(basePackages = "com.vielheit.graph.repository")
@ComponentScan({"com.vielheit.graph"})
public class Neo4jConfig {
    @Bean
    public SessionFactory getSessionFactory() {
        return new SessionFactory("com.vielheit.graph.domain");
    }

    @Bean(name="graphTransactionManager")
    public Neo4jTransactionManager transactionManager() {
        return new Neo4jTransactionManager(getSessionFactory());
    }

    @Bean
    @ConfigurationProperties(prefix="spring.data.neo4j")
    public DataSource secondaryDataSource() {
        return DataSourceBuilder.create().build();
    }
}

File: src/main/kotlin/org/vielheit/core/config/Neo4jConfig.kt
package org.vielheit.core.config

import org.neo4j.ogm.session.SessionFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.neo4j.repository.config.EnableNeo4jRepositories
import org.springframework.data.neo4j.transaction.Neo4jTransactionManager
import org.springframework.transaction.annotation.EnableTransactionManagement
import org.springframework.boot.jdbc.DataSourceBuilder
import org.springframework.boot.context.properties.ConfigurationProperties
import org.vielheit.core.config.properties.Neo4jProperties
import org.vielheit.core.util.logger
import javax.sql.DataSource


@Configuration
@EnableTransactionManagement
@EnableNeo4jRepositories("org.vielheit.core.repository")
class Neo4jConfig(val props: Neo4jProperties) {
    val log by logger()

    init { log.info("Using neo4j {}", props.uri) }

    @Bean
    fun sessionFactory(): SessionFactory = SessionFactory(configuration(), "org.vielheit.core.domain")

    @Bean("graphTxManager")
    fun transactionManager(): Neo4jTransactionManager = Neo4jTransactionManager(sessionFactory())

    @Bean
    @ConfigurationProperties(prefix = "vielheit.neo4j")
    fun secondaryDataSource(): DataSource = DataSourceBuilder.create().build()

    @Bean
    fun configuration(): org.neo4j.ogm.config.Configuration = org.neo4j.ogm.config.Configuration.Builder()
            .uri(props.uri)
            .credentials(props.username, props.password)
            .autoIndex("assert")
            .build()
}